kernel/userspace.c

```diff
@@ -130,41 +149,56 @@ uint8_t *z_priv_stack_find(k_thread_stack_t *stack)
 #define DYN_OBJ_DATA_ALIGN_K_THREAD	(sizeof(void *))
 #endif
 
+#ifdef CONFIG_DYNAMIC_THREAD_STACK_SIZE
+#ifndef CONFIG_MPU_STACK_GUARD
+#define DYN_OBJ_DATA_ALIGN_K_THREAD_STACK \
+	Z_THREAD_STACK_OBJ_ALIGN(CONFIG_PRIVILEGED_STACK_SIZE)
+#else
+#define DYN_OBJ_DATA_ALIGN_K_THREAD_STACK \
+	Z_THREAD_STACK_OBJ_ALIGN(CONFIG_DYNAMIC_THREAD_STACK_SIZE)
+#endif /* !CONFIG_MPU_STACK_GUARD */
+#else
+#define DYN_OBJ_DATA_ALIGN_K_THREAD_STACK \
+	Z_THREAD_STACK_OBJ_ALIGN(ARCH_STACK_PTR_ALIGN)
+#endif /* CONFIG_DYNAMIC_THREAD_STACK_SIZE */
+
 #define DYN_OBJ_DATA_ALIGN		\
 	MAX(DYN_OBJ_DATA_ALIGN_K_THREAD, (sizeof(void *)))
 
-struct dyn_obj {
+struct dyn_obj_base {
 	struct z_object kobj;
-	sys_dnode_t dobj_list;
-	struct rbnode node; /* must be immediately before data member */
+	sys_dnode_t dobj_list; /* must be immediately before data member */
+};
+
+struct dyn_obj {
+	struct dyn_obj_base base;
 
 	/* The object itself */
 	uint8_t data[] __aligned(DYN_OBJ_DATA_ALIGN_K_THREAD);
 };
 
+/* Thread stacks impose a very restrict alignment. Use this alignment
+ * (generally page size) for all objects will cause a lot waste memory.
+ */
+struct dyn_obj_stack {
+	struct dyn_obj_base base;
+
+	/* The object itself */
+	uint8_t data[] __aligned(DYN_OBJ_DATA_ALIGN_K_THREAD_STACK);
```

Comment:
yep ... there is this waste of memory indeed, that is probably no longer needed now that I am searching the object linearly. I can now do two allocations and avoid it. Let met try it out.

Reply:
I removed this alignment requirement for the object and make a new allocation for the stack data itself (with the proper alignment), this can potentially avoid some memory waste.